8Knot/pages/codebase/visualizations/cntrb_file_heatmap.py

```diff
@@ -0,0 +1,374 @@
+from dash import html, dcc, callback
+import dash
+import dash_bootstrap_components as dbc
+import dash_mantine_components as dmc
+from dash.dependencies import Input, Output, State
+import plotly.graph_objects as go
+import pandas as pd
+import logging
+from dateutil.relativedelta import *  # type: ignore
+import plotly.express as px
+from pages.utils.graph_utils import get_graph_time_values, color_seq
+from queries.contributors_query import contributors_query as cnq
+from queries.cntrb_per_file_query import cntrb_per_file_query as cpfq
+from queries.repo_files_query import repo_files_query as rfq
+from app import augur
+import io
+from cache_manager.cache_manager import CacheManager as cm
+from pages.utils.job_utils import nodata_graph
+import time
+from dash.exceptions import PreventUpdate
+
+PAGE = "codebase"
+VIZ_ID = "cntrb-file-heatmap"
+
+# div to hold all objects to wait for loading to render
+graph_loading = html.Div(
+    [
+        dbc.Popover(
+            [
+                dbc.PopoverHeader("Graph Info:"),
+                dbc.PopoverBody("INSERT CONTEXT OF GRAPH HERE"),
```

Comment:
Need to workshop this together in PR comments

Reply:
Agreed, lets discuss more. Also think it might be a good idea to get a "good enough" option for a first round pass, open an issue for discussion, and get feedback from people once its hosted on 8Knot

Reply:
Yeah that's just fine with me. Doesn't need to be perfect, just has to effectively communicate a reasonable interpretation